### PR TITLE
feat: update notebook server images

### PR DESCRIPTION
**Pull request by thesuperzapper**
This PR updates all the `example-notebook-servers` images.

The key thing this does is pin `h5py==3.10.0` in the `jupyter-tensorflow` image, because this was causing builds to fail for ARM64, as `h5py` has stopped shipping wheels for ARM64 in `3.11.0` (https://github.com/h5py/h5py/issues/2408).

Other significant changes:

- Updated PyTorch to `2.3.0`
- Updated TensorFlow to `2.15.1`
- Updated code-server to: `4.89.1`
- Updated JupyterLab to: `4.2.1`
- Updated Jupyter to `7.2.0`
- Updated `kfp` SDK to: `2.7.0`
- Moved the `s6/cont-init.d/01-copy-tmp-home` script to only be on the `base` image, because it was not necessary to have the same code copied across multiple images.

Finally, I have added docs about building the images locally based on feedback from https://github.com/kubeflow/kubeflow/issues/7587.

### CODE REVIEW

**Comment by thesuperzapper** [comment]
I am getting a strange error on the `jupyter-tensorflow` images, which we will need to figure out before merging.

Its probably something simple like that I increased the `h5py` version too much, and should pin it to a lower version.

Specifically, I am getting this error:

```
[I 2024-05-27 04:07:50.672 ServerApp](B AsyncIOLoopKernelRestarter: restarting kernel (5/5), new random ports
2024-05-27T04:07:50.860691588Z Traceback (most recent call last):
2024-05-27T04:07:50.860720671Z   File "<frozen runpy>", line 198, in _run_module_as_main
2024-05-27T04:07:50.860722838Z   File "<frozen runpy>", line 88, in _run_code
2024-05-27T04:07:50.860724088Z   File "/opt/conda/lib/python3.11/site-packages/ipykernel_launcher.py", line 16, in <module>
2024-05-27T04:07:50.860725504Z     from ipykernel import kernelapp as app
2024-05-27T04:07:50.860726463Z   File "/opt/conda/lib/python3.11/site-packages/ipykernel/kernelapp.py", line 21, in <module>
2024-05-27T04:07:50.860727463Z     from IPython.core.application import (  # type:ignore[attr-defined]
2024-05-27T04:07:50.860728379Z   File "/opt/conda/lib/python3.11/site-packages/IPython/__init__.py", line 55, in <module>
2024-05-27T04:07:50.860729421Z     from .terminal.embed import embed
2024-05-27T04:07:50.860730296Z   File "/opt/conda/lib/python3.11/site-packages/IPython/terminal/embed.py", line 16, in <module>
2024-05-27T04:07:50.860731296Z     from IPython.terminal.interactiveshell import TerminalInteractiveShell
2024-05-27T04:07:50.860732171Z   File "/opt/conda/lib/python3.11/site-packages/IPython/terminal/interactiveshell.py", line 48, in <module>
2024-05-27T04:07:50.860733171Z     from .debugger import TerminalPdb, Pdb
2024-05-27T04:07:50.860734046Z   File "/opt/conda/lib/python3.11/site-packages/IPython/terminal/debugger.py", line 6, in <module>
2024-05-27T04:07:50.860773671Z     from IPython.core.completer import IPCompleter
2024-05-27T04:07:50.860785171Z   File "/opt/conda/lib/python3.11/site-packages/IPython/core/completer.py", line 219, in <module>
2024-05-27T04:07:50.861306963Z     from IPython.core.guarded_eval import guarded_eval, EvaluationContext
2024-05-27T04:07:50.861315546Z   File "/opt/conda/lib/python3.11/site-packages/IPython/core/guarded_eval.py", line 40, in <module>
2024-05-27T04:07:50.861317421Z     from typing_extensions import TypeAliasType
2024-05-27T04:07:50.861318838Z ImportError: cannot import name 'TypeAliasType' from 'typing_extensions' (/opt/conda/lib/python3.11/site-packages/typing_extensions.py)
```

**Comment by thesuperzapper** [comment]
@kimwnasptd this should be ready to review now, all images work in my testing for ARM/X86.

And the CUDA images work with my GPU tests (NOTE: we only ship CUDA on X86, and ARM does not work for them).

File: components/example-notebook-servers/jupyter-pytorch-full/requirements.txt [hunk 1]
```diff
@@ -1,6 +1,5 @@
 # kubeflow packages
-kfp==2.4.0
-kfp-server-api==2.0.3
+kfp==2.7.0
```
Comment (andreyvelich) [inline components/example-notebook-servers/jupyter-pytorch-full/requirements.txt:2]: @kubeflow/wg-notebooks-leads @kubeflow/wg-training-leads Can we install [`kubeflow-training`](https://pypi.org/project/kubeflow-training/) and [`kubeflow-katib`](https://pypi.org/project/kubeflow-katib/) Python SDK as part of Kubeflow Notebooks images ?
Similar to KFP ?

Comment (thesuperzapper) [reply components/example-notebook-servers/jupyter-pytorch-full/requirements.txt:2]: Is there a particular version of these packages that aligns with the 1.9 release?

Comment (andreyvelich) [reply components/example-notebook-servers/jupyter-pytorch-full/requirements.txt:2]: Could we include the `kubeflow-training==1.8.0rc0` and `kubeflow-katib==0.17.0rc0` ?
We haven't released the 1.8 and 0.17 version yet.

Comment (thesuperzapper) [reply components/example-notebook-servers/jupyter-pytorch-full/requirements.txt:2]: @andreyvelich lets wait until the final versions of those packages are released.

Can you try and cut them soon, so we can include them in the next RC?

Adding packages will require me to run all the tests again for all the images, which I don't have time to do before we cut the first RC for 1.9.0.

**Comment by kimwnasptd** [comment]
Agree with @thesuperzapper to proceed with this PR for now to cut our first RC. And then let's include the rest in the following RCs

/lgtm
/approve

**Pull request merged**